### PR TITLE
Add a progress bar for cli based hashing

### PR DESCRIPTION
**Pull request by hardbyte**
Adds a simple progress bar for the command line utility.

screenshot from 2017-12-13 14-35-54

I think the `generate_clk_from_csv` function should expose a better api for progress for other callers though - maybe a progress queue, or callback function?

### CODE REVIEW

File: clkhash/clk.py [hunk 1]
```diff
@@ -70,6 +72,22 @@ def generate_clk_from_csv(input, keys, schema_types, no_header=False):
                                          chunk, schema_types, key_lists)
                 futures.append(future)
 
+            if progress_bar:
+                future_iterator = tqdm(concurrent.futures.as_completed(futures),
+                                       desc="Hashing",
+                                       total=len(pii_data) // chunk_size,
+                                       unit='KH',
```
Comment (wilko77) [inline clkhash/clk.py]: this unit is only true if `chunk_size = 1000`

File: clkhash/clk.py [hunk 2]
```diff
@@ -70,6 +72,22 @@ def generate_clk_from_csv(input, keys, schema_types, no_header=False):
                                          chunk, schema_types, key_lists)
                 futures.append(future)
 
+            if progress_bar:
+                future_iterator = tqdm(concurrent.futures.as_completed(futures),
+                                       desc="Hashing",
+                                       total=len(pii_data) // chunk_size,
```
Comment (wilko77) [inline clkhash/clk.py]: you will miss the last chunk if `len(pii_data)` is not a multiple of `chunk_size`

File: clkhash/clk.py [hunk 3]
```diff
@@ -70,6 +72,22 @@ def generate_clk_from_csv(input, keys, schema_types, no_header=False):
...
+                                       leave=False,
+                                       )
+            else:
+                future_iterator = concurrent.futures.as_completed(futures)
+
+            for f in future_iterator:
+                # Note these are not in order of submission - but completion!
+                # We do this noop loop inorder to deal with progress as the futures
+                # complete
+                pass
```
Comment (wilko77) [inline clkhash/clk.py]: This is not very elegant. Now you iterate twice over the futures. And this loop of `pass`.... tststs.... 
Have a loop at my first [stab](https://gist.github.com/wilko77/bba9580ca8aa5b9c98b403002fba7634). I particularly like the `add_done_callback` on futures. 
Also, you don't report progress for the single thread version.

File: clkhash/clk.py (hunk 2 above)
Comment (hardbyte) [reply clkhash/clk.py]: 😳

File: clkhash/clk.py (hunk 1 above)
Comment (hardbyte) [reply clkhash/clk.py]: Ahh, I did change the chunk size to 1000 I swear!

Then I forgot why and changed it back...

File: clkhash/clk.py (hunk 3 above)
Comment (hardbyte) [reply clkhash/clk.py]: I agree your implementation is much better - I've pushed your changes as a commit and then tweaked the callback to be a bit more general.

**Comment by hardbyte** [comment]
Thanks for the code review Wilko, I think I've addresses all your requested changes.

The only piece I'm not happy with in the `generate_clks` function is we should probably split by "executor" and leave it up to the user to decide - and only go back to the if Python version > 3 decision if the user didn't provide a preference. There is of course no reason that Python2 users can't have parallel execution either - I just wanted to use the nicer concurrent api when it exists for Python 3.

**Pull request merged**